xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-templates/src/main/resources/saverights.vm

```diff
@@ -28,8 +28,15 @@
 
 #if ($!{services.csrf.isTokenValid("$!{request.getParameter('form_token')}")})
 #foreach($obj in $doc.getObjects($clsname)) ## XWiki.XWikiGlobalRights or XWiki.XWikiRights
+  ## In subwikis, users or groups may be referenced by "wiki:Space.Username" or "Space.Username"
+  #set ($wikiname = $xcontext.database)
+  #if ($wikiname == "xwiki")
+    #set ($uorgmatch = "${regextool.quote($fullname)}")
+  #else
+    #set ($uorgmatch = "(" + ${regextool.quote($wikiname)} + ":)?" + "${regextool.quote($fullname)}")
+  #end
   #set($pers = "$!obj.getProperty($uorg).getValue()")
-  #if($pers.matches("${regextool.quote($fullname)},?")) ## found the rule containing that person
+  #if($pers.matches("${uorgmatch},?")) ## found the rule containing that person
```

Comment:
What happens if there are two users ``XWiki.Bob`` and ``XWiki.Bobby``? Using regular expressions for this feels error-prone to me. Why not resolve each user/group reference from the value (relative to the current wiki) and compare the absolute reference?

Reply:
I agree with you that using references is more appropriate and will do so. I just checked the case you mention to make sure the existing product code (which is based on regular expressions) was not misconfiguring access rights in such a case and it is ok, both with single and multiple user(s)/group(s) rules, because of the following:

```
#set ($pers = 'XWiki.bobby')
#set ($regex = $regextool.quote('XWiki.bob'))
#set ($result = $pers.matches("$regex,?")) ## Does not match. Would match with "$regex,?.*"

#set ($pers = 'XWiki.bobby,XWiki.alice')
#set ($result = $pers.matches("^(.*,)?$regex(,.*)?$")) ## Does not match

#set ($pers = 'XWiki.bobby,XWiki.bob,XWiki.alice')
#set ($result = $pers.matches("^(.*,)?$regex(,.*)?$")) ## Matches, however the subsequent replace calls do not replace anything:
#set ($pers = $pers.replaceAll("^${regex}(,|$)", "")) ## No match because the regex needs to be followed by a comma, a pipe, or end of line
#set ($pers = $pers.replaceAll("(^|,)${regex}(,|$)", "$2")) ## No match either for the same reason
```

Reply:
dce6e16 uses references instead of regular expressions. I tested the code with single and multiple user/group access rules on both a subwiki and the main wiki with relative or absolute references. There might be shorthand methods for dealing in a uniform manner with property values consisting of strings separated with comma or pipes but couldn't find any so far. 

Other templates use regular expressions for matching a given user or group identifier, eg `rightsUI.vm`, `getusersandgroups.vm`, I'll create issues for using references in these templates as well.

Reply:
The latest change in this pull request resolves each user/group into a reference indeed.